hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java

```diff
@@ -100,14 +102,16 @@ private boolean fetchPage(
         int cursorBufferSize,
         InternalSerializationService serializationService
     ) {
+        long endTime = System.nanoTime() + SECONDS.toNanos(1);
         while (iterator.hasNext()) {
             SqlRow row = iterator.next();
             Row rowInternal = ((SqlRowImpl) row).getDelegate();
             Data rowData = serializationService.toData(rowInternal);
 
             page.add(rowData);
 
-            if (page.size() == cursorBufferSize) {
+            // TODO we call nanoTime for each item - use batching
+            if (page.size() == cursorBufferSize || System.nanoTime() >= endTime) {
```

Comment:
1) Isn't System.currentTimeMillis enough here? (though, I do not know any performance difference between `nano` and `millis`
2) The last line in this method calls `hasNext` which is blocking as well. Therefore, the idea of early return doesn't really work in this prototype. Instead, the method should return as soon as the very first call to `hasNext` returned something

Reply:
`nanoTime()` should always be used for elapsed time. CTM isn't monotonic in rare cases, doesn't run at constant speed. Performance differs by OS and JVM implementation, but I think on Linux they are comparable. IIRC on Windows CTM is faster, but also much less precise (~15ms resolution)

Reply:
problem #2: fixed